# Restoring gold from ore

## Description of the project

We will prepare a prototype of machine learning model for “numbers”.The company is developing solutions for the effective operation of industrial enterprises.
The model should predict the coefficient of restoration of gold from gold -containing ore.At our disposal, data with production and cleaning parameters.
The model will help optimize production so as not to launch an enterprise with unprofitable characteristics.

We need:
- prepare data;
- conduct a research analysis of data;
- Build and train the model.

The technological process of mining from ore.We will carefully study the stages of the process.

When the mined ore undergoes primary processing, the crushed mixture is obtained.It is sent to flotation (enrichment) and two -stage cleaning.

<IMG SRC = "https://pictures.s3.yandex.net/resources/viruchka_1576238830.jpg"
Alt = "Stages of processing"
Width = "800"
>

We describe each stage:
1. Flotation
A mixture of gold -containing ore is supplied to the flotation installation.After enrichment, a black concentrate and “dump tails” are obtained, that is, the remains of a product with a low concentration of valuable metals.
The stability of this process is affected by the unstable and non-optimal physico-chemical state of the flotation pulp (a mixture of solid particles and liquid).
2. Cleaning
Black concentrate passes two cleaning.The output is final concentrate and new dump tails.

Data description

Process
- `Rougher feed` - The starting raw material
- `Rougher Additions` (or` Reagent Additions`) - Flotation reagents: `Xanthate`,` Sulphate`, `Depressant`
- `xanthate` ** - xanthogenate (promoter, or flotation activator);
- `sulphate` - sulfate (in this production of sodium sulfide);
- `Depressant` - Depressant (sodium silicate).
- `Rougher Process` (English" rude process ") - flotation
- `Rougher Tails' - Dumped tails
- `Float Banks` - Flotation installation
- `Cleaner Process` - Cleaning
- `rougher au` - black gold concentrate
- `final au` - final gold concentrate

Parameters of stages
- `Air amount` - Air volume
- `fluid levels` - fluid level
- `fed size` - the size of granules of raw materials
- `fed rate` - feed speed

The name of the signs

The name of the signs should be this:
`[stage]. [Type_paeter]. [name_paeter]`

Example: `rogher.input.feed_ag`

Possible values ​​for the block [Stage]:
- `rogher` - flotation
- `primary_cleaner` - primary cleaning
- `secondary_cleaner` - secondary cleaning
- `final` - final characteristics

Possible values ​​for the block [type_paeter]:
- `Input` - raw materials parameters
- `output` - product parameters
- `state - parameters characterizing the current state of the stage
- `Calculation` - calculated characteristics

Calculation of efficiency

We need to simulate the process of restoring gold from gold -containing ore.

The effectiveness of enrichment is calculated by the formula:

$ $ Recovery = \ frac {c * (f - t)} {f * (c - t)} $ $

Where:
- `c` - the share of gold in the concentrate after flotation/cleaning;
- `f` - the share of gold in raw materials/concentrate to flotation/cleaning;
- `t` - the share of gold in the dash tails after flotation/cleaning.

The metric of quality

To solve the problem, we introduce a new quality metric - `Smape` (English Symmetric Mean Absolute Percentage Error,“ Symmetric average absolute percentage deviation ”).
It is similar to `mae`, but it is not expressed in absolute values, but in relative ones.Why symmetric?It equally takes into account the scale of both the target feature and predictions.
Smape metric is calculated as follows:

$ {smape} = \ frac {1} {n} \ sum_ {i = 1}^{n} \ frac {| y_i - \ hat {y_i} {(| y_i | + | \ hat {y_i}}}|) / 2} * 100 \% $ $

Designations:

- $ y_i $ the value of the target feature for an object with serial number I in the sample on which quality is measured.

- $ \ HAT {y_i} $ the value of prediction for an object with ordinal number I, for example, in a test sample.

- $ n $ the number of objects in the sample.

- $ \ sum_ {i = 1}^{n} $ Summing for all sampling objects (I changes from 1 to N).

It is necessary to predict two values ​​at once:
- the effectiveness of the enrichment of the draft concentrate `rougher.autPut.recovery`;
- The effectiveness of enrichment of the final concentrate `final.autPut.recovery`.

The final metric consists of two quantities:

$ $ {Final Smape} = 25 \% * {Smape (Rougher)} + ​​75 \% * {SMAPE (FINAL)} $ dollar

Data are in three files:
- `gold_recovery_train_new.csv` - training sample;
- `gold_recovery_test_new.csv` - test sample;
- `gold_recovery_full_new.csv` - initial data.

Data is indexed by the date and time of receipt of information (sign `Date`).Time neighboring parameters are often similar.

Some parameters are not available, because they are measured and/or calculated much later.Because of this, there are no signs in the test sample that may be in the training.There are also no targeted signs in the test set.

The initial dataset contains training and test samples with all the signs.
Raw data at our disposal: they were simply unloaded from the storage.Before proceeding with the construction of the model, we will check them according to our instructions for correctness.

- 1. Prepare data
- 1.1.Open the files and study them.
- Way to files:
- `/datasets/gold_recovery_train_new.csv.` [download Dataset] (https://code.s3.s3.yandex.net/datasets/gold_recovery_train_new.csv" gold_recovery_train_new.csv ")
- `/datasets/gold_recovery_test_new.csv.` [download Dataset] (https://code.s3.yandex.net/datasets/gold_recovery_test_new.csv"
- `/datasets/gold_recovery_full_new.csv.` [download Dataset] (https://code.s3.s3.yandex.net/datasets/gold_recovery_full_new.csv" gold_recovery_full_new.csv ")
- 1.2.We check that the effectiveness of enrichment is calculated correctly.We calculate it on the training sample for a sign `rogher.Ontput.recovery`.Find `mae` between your calculations and the value of the sign.We describe the conclusions.
- 1.3.Let us analyze the signs inaccessible in the test sample.
- 1.4.We will exploit data.
- 2. We analyze the data
- 2.1.Let's see how the concentration of metals (AU, AG, PB) changes at various stages of cleaning.Opening conclusions.
- 2.2.Compare the distribution of the size of the granules of raw materials in the training and test samples.(If the distributions are very different from each other, the evaluation of the model will be wrong).
- 2.3.We examine the total concentration of all substances at different stages: in raw materials, in draft and final concentrates.
- 3. We will build a model
- 3.1.We will write a function for calculating the final `smape`.
- 3.2.We will train different models and evaluate their quality cross-nineteration.We choose the best model and check it in a test sample.We describe the conclusions.

## Data Preparation

### Loading and Inspection

In [2]:
# отключим предупреждающие уведомления
import warnings
warnings.filterwarnings('ignore')

# Добавим основные библиотеки
import pandas as pd
import numpy as np
import os.path
import math
 
# Установим и настроим plotly и необходимые компоненты для отображения графиков
# !pip install chart-studio
# !pip install plotly
# !pip install cufflinks

from chart_studio import plotly
from plotly.offline import iplot
import cufflinks as cf
cf.go_offline()
cf.set_config_file(world_readable=True, theme='henanigans', offline=True)

import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('dark_background')

# установим необходимые компониенты из библиотеки ылдуфкт для построения моделей
from sklearn.metrics import mean_absolute_error, mean_squared_error, make_scorer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score, KFold, RandomizedSearchCV
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

# Сделаем небольшие настройки pandas для комфортного отображения проекта
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
pd.options.mode.chained_assignment = None

In [3]:
# Импортируем датасеты методом try/except и установим индексом столбец date
try:
    path_train = os.path.join('datasets/gold_recovery_train_new.csv', index_col='date')
    train = pd.read_csv(path_train)
except:
    train = pd.read_csv('https://code.s3.yandex.net/datasets/gold_recovery_train_new.csv', index_col='date')

train = train.set_index(pd.DatetimeIndex(train.index))

try:
    path_test = os.path.join('datasets/gold_recovery_test_new.csv', index_col='date')
    test = pd.read_csv(path_test)
except:
    test = pd.read_csv('https://code.s3.yandex.net/datasets/gold_recovery_test_new.csv', index_col='date')

test = test.set_index(pd.DatetimeIndex(test.index))

try:
    path_full = os.path.join('datasets/gold_recovery_full_new.csv', index_col='date')
    full = pd.read_csv(path_full)
except:
    full = pd.read_csv('https://code.s3.yandex.net/datasets/gold_recovery_full_new.csv', index_col='date')

full = full.set_index(pd.DatetimeIndex(full.index))

In [4]:
# Функция для вывода первичной информации по датасету
def first_view(data):
    title_lenght = 79
    title_one = 'Обзор данных первых и последних пяти строк'
    print('~' * int((title_lenght - len(title_one)) / 2), title_one, '~' * (title_lenght - len(title_one) - int((title_lenght - len(title_one)) / 2)))
    display(data.head(5))
    display(data.tail(5))
    print()
    title_two = 'Основная описательная статистика'
    print('~' * int((title_lenght - len(title_two)) / 2), title_two, '~' * (title_lenght - len(title_two) - int((title_lenght - len(title_two)) / 2)))
    display(data.describe())
    print()
    title_three = 'Тип данных' 
    print('~' * int((title_lenght - len(title_three)) / 2), title_three, '~' * (title_lenght - len(title_three) - int((title_lenght - len(title_three)) / 2)))
    data.info()

In [5]:
first_view(train)

~~~~~~~~~~~~~~~~~~ Обзор данных первых и последних пяти строк ~~~~~~~~~~~~~~~~~~~


final.output.concentrate_ag  final.output.concentrate_pb  \
date                                                                            
2016-01-15 00:00:00                         6.06                         9.89   
2016-01-15 01:00:00                         6.03                         9.97   
2016-01-15 02:00:00                         6.06                        10.21   
2016-01-15 03:00:00                         6.05                         9.98   
2016-01-15 04:00:00                         6.15                        10.14   

                     final.output.concentrate_sol  \
date                                                
2016-01-15 00:00:00                          5.51   
2016-01-15 01:00:00                          5.26   
2016-01-15 02:00:00                          5.38   
2016-01-15 03:00:00                          4.86   
2016-01-15 04:00:00                          4.94   

                     final.output.concentrate_au  final.output.recovery  \
date                                                                      
2016-01-15 00:00:00                        42.19                  70.54   
2016-01-15 01:00:00                        42.70                  69.27   
2016-01-15 02:00:00                        42.66                  68.12   
2016-01-15 03:00:00                        42.69                  68.35   
2016-01-15 04:00:00                        42.77                  66.93   

                     final.output.tail_ag  final.output.tail_pb  \
date                                                              
2016-01-15 00:00:00                 10.41                  0.90   
2016-01-15 01:00:00                 10.46                  0.93   
2016-01-15 02:00:00                 10.51                  0.95   
2016-01-15 03:00:00                 10.42                  0.88   
2016-01-15 04:00:00                 10.36                  0.79   

                     final.output.tail_sol  final.output.tail_au  \
date                                                               
2016-01-15 00:00:00                  16.90                  2.14   
2016-01-15 01:00:00                  16.63                  2.22   
2016-01-15 02:00:00                  16.21                  2.26   
2016-01-15 03:00:00                  16.53                  2.15   
2016-01-15 04:00:00                  16.53                  2.06   

                     primary_cleaner.input.sulfate  \
date                                                 
2016-01-15 00:00:00                         127.09   
2016-01-15 01:00:00                         125.63   
2016-01-15 02:00:00                         123.82   
2016-01-15 03:00:00                         122.27   
2016-01-15 04:00:00                         117.99   

                     primary_cleaner.input.depressant  \
date                                                    
2016-01-15 00:00:00                             10.13   
2016-01-15 01:00:00                             10.30   
2016-01-15 02:00:00                             11.32   
2016-01-15 03:00:00                             11.32   
2016-01-15 04:00:00                             11.91   

                     primary_cleaner.input.feed_size  \
date                                                   
2016-01-15 00:00:00                             7.25   
2016-01-15 01:00:00                             7.25   
2016-01-15 02:00:00                             7.25   
2016-01-15 03:00:00                             7.25   
2016-01-15 04:00:00                             7.25   

                     primary_cleaner.input.xanthate  \
date                                                  
2016-01-15 00:00:00                            0.99   
2016-01-15 01:00:00                            1.00   
2016-01-15 02:00:00                            0.99   
2016-01-15 03:00:00                            1.00   
2016-01-15 04:00:00                            1.01   

                     primary_cleaner.output.concen

final.output.concentrate_ag  final.output.concentrate_pb  \
date                                                                            
2018-08-18 06:59:59                         3.22                        11.36   
2018-08-18 07:59:59                         3.20                        11.35   
2018-08-18 08:59:59                         3.11                        11.43   
2018-08-18 09:59:59                         3.37                        11.63   
2018-08-18 10:59:59                         3.60                        11.74   

                     final.output.concentrate_sol  \
date                                                
2018-08-18 06:59:59                          6.80   
2018-08-18 07:59:59                          6.86   
2018-08-18 08:59:59                          6.89   
2018-08-18 09:59:59                          6.80   
2018-08-18 10:59:59                          6.72   

                     final.output.concentrate_au  final.output.recovery  \
date                                                                      
2018-08-18 06:59:59                        46.71                  73.76   
2018-08-18 07:59:59                        46.87                  69.05   
2018-08-18 08:59:59                        46.80                  67.00   
2018-08-18 09:59:59                        46.41                  65.52   
2018-08-18 10:59:59                        46.30                  70.28   

                     final.output.tail_ag  final.output.tail_pb  \
date                                                              
2018-08-18 06:59:59                  8.77                  3.14   
2018-08-18 07:59:59                  8.90                  3.13   
2018-08-18 08:59:59                  8.53                  2.91   
2018-08-18 09:59:59                  8.78                  2.82   
2018-08-18 10:59:59                  8.41                  2.52   

                     final.output.tail_sol  final.output.tail_au  \
date                                                               
2018-08-18 06:59:59                  10.40                  1.53   
2018-08-18 07:59:59                  10.55                  1.61   
2018-08-18 08:59:59                  11.12                  1.60   
2018-08-18 09:59:59                  10.46                  1.60   
2018-08-18 10:59:59                  10.65                  1.39   

                     primary_cleaner.input.sulfate  \
date                                                 
2018-08-18 06:59:59                         123.38   
2018-08-18 07:59:59                         120.88   
2018-08-18 08:59:59                         105.67   
2018-08-18 09:59:59                          98.88   
2018-08-18 10:59:59                          95.25   

                     primary_cleaner.input.depressant  \
date                                                    
2018-08-18 06:59:59                              8.03   
2018-08-18 07:59:59                              7.96   
2018-08-18 08:59:59                              7.96   
2018-08-18 09:59:59                              7.98   
2018-08-18 10:59:59                              8.08   

                     primary_cleaner.input.feed_size  \
date                                                   
2018-08-18 06:59:59                             6.50   
2018-08-18 07:59:59                             6.50   
2018-08-18 08:59:59                             6.50   
2018-08-18 09:59:59                             6.50   
2018-08-18 10:59:59                             6.50   

                     primary_cleaner.input.xanthate  \
date                                                  
2018-08-18 06:59:59                            1.30   
2018-08-18 07:59:59                            1.30   
2018-08-18 08:59:59                            1.32   
2018-08-18 09:59:59                            1.24   
2018-08-18 10:59:59                            1.28   

                     primary_cleaner.output.concen


~~~~~~~~~~~~~~~~~~~~~~~ Основная описательная статистика ~~~~~~~~~~~~~~~~~~~~~~~~


final.output.concentrate_ag  final.output.concentrate_pb  \
count                     14148.00                     14148.00   
mean                          5.14                        10.13   
std                           1.37                         1.65   
min                           0.00                         0.00   
25%                           4.21                         9.30   
50%                           4.99                        10.30   
75%                           5.86                        11.17   
max                          16.00                        17.03   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                      13938.00                     14149.00   
mean                           9.20                        44.00   
std                            2.79                         4.91   
min                            0.00                         0.00   
25%                            7.48                        43.28   
50%                            8.85                        44.87   
75%                           10.49                        46.17   
max                           18.12                        52.76   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count               14149.00              14149.00              14049.00   
mean                   66.52                  9.61                  2.60   
std                    10.30                  2.32                  0.97   
min                     0.00                  0.00                  0.00   
25%                    62.55                  8.00                  1.91   
50%                    67.43                  9.48                  2.59   
75%                    72.35                 11.00                  3.24   
max                   100.00                 19.55                  5.64   

       final.output.tail_sol  final.output.tail_au  \
count               14144.00              14149.00   
mean                   10.51                  2.92   
std                     3.00                  0.90   
min                     0.00                  0.00   
25%                     8.81                  2.37   
50%                    10.51                  2.85   
75%                    11.93                  3.43   
max                    22.32                  8.20   

       primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                       14129.00                          14117.00   
mean                          133.32                              8.87   
std                            39.43                              3.36   
min                             0.00                              0.00   
25%                           107.01                              6.04   
50%                           133.02                              8.04   
75%                           159.83                             11.52   
max                           250.13                             20.05   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                         14149.00                        14049.00   
mean                              7.32                            0.89   
std                               0.61                            0.37   
min                               1.08                            0.00   
25%                               6.96                            0.61   
50%                               7.29                            0.89   
75%                               7.70                            1.10   
max                              10.47                            2.51   

       primary_cleaner.output.concentrate_ag  \
count                               14149.00   
mean                                    8.20   
std                                     2.01   
min                                     0.00   
25%                                     7.11


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Тип данных ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 14149 entries, 2016-01-15 00:00:00 to 2018-08-18 10:59:59
Data columns (total 86 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   final.output.concentrate_ag                         14148 non-null  float64
 1   final.output.concentrate_pb                         14148 non-null  float64
 2   final.output.concentrate_sol                        13938 non-null  float64
 3   final.output.concentrate_au                         14149 non-null  float64
 4   final.output.recovery                               14149 non-null  float64
 5   final.output.tail_ag                                14149 non-null  float64
 6   final.output.tail_pb                                14049 non-null  float64
 7   final.output.tail_sol                  

In [6]:
first_view(test)

~~~~~~~~~~~~~~~~~~ Обзор данных первых и последних пяти строк ~~~~~~~~~~~~~~~~~~~


primary_cleaner.input.sulfate  \
date                                                 
2016-09-01 00:59:59                         210.80   
2016-09-01 01:59:59                         215.39   
2016-09-01 02:59:59                         215.26   
2016-09-01 03:59:59                         215.34   
2016-09-01 04:59:59                         199.10   

                     primary_cleaner.input.depressant  \
date                                                    
2016-09-01 00:59:59                             14.99   
2016-09-01 01:59:59                             14.99   
2016-09-01 02:59:59                             12.88   
2016-09-01 03:59:59                             12.01   
2016-09-01 04:59:59                             10.68   

                     primary_cleaner.input.feed_size  \
date                                                   
2016-09-01 00:59:59                             8.08   
2016-09-01 01:59:59                             8.08   
2016-09-01 02:59:59                             7.79   
2016-09-01 03:59:59                             7.64   
2016-09-01 04:59:59                             7.53   

                     primary_cleaner.input.xanthate  \
date                                                  
2016-09-01 00:59:59                            1.01   
2016-09-01 01:59:59                            0.99   
2016-09-01 02:59:59                            1.00   
2016-09-01 03:59:59                            0.86   
2016-09-01 04:59:59                            0.81   

                     primary_cleaner.state.floatbank8_a_air  \
date                                                          
2016-09-01 00:59:59                                 1398.98   
2016-09-01 01:59:59                                 1398.78   
2016-09-01 02:59:59                                 1398.49   
2016-09-01 03:59:59                                 1399.62   
2016-09-01 04:59:59                                 1401.27   

                     primary_cleaner.state.floatbank8_a_level  \
date                                                            
2016-09-01 00:59:59                                   -500.23   
2016-09-01 01:59:59                                   -500.06   
2016-09-01 02:59:59                                   -500.87   
2016-09-01 03:59:59                                   -498.86   
2016-09-01 04:59:59                                   -500.81   

                     primary_cleaner.state.floatbank8_b_air  \
date                                                          
2016-09-01 00:59:59                                 1399.14   
2016-09-01 01:59:59                                 1398.06   
2016-09-01 02:59:59                                 1398.86   
2016-09-01 03:59:59                                 1397.44   
2016-09-01 04:59:59                                 1398.13   

                     primary_cleaner.state.floatbank8_b_level  \
date                                                            
2016-09-01 00:59:59                                   -499.92   
2016-09-01 01:59:59                                   -499.78   
2016-09-01 02:59:59                                   -499.76   
2016-09-01 03:59:59                                   -499.21   
2016-09-01 04:59:59                                   -499.50   

                     primary_cleaner.state.floatbank8_c_air  \
date                                                          
2016-09-01 00:59:59                                 1400.10   
2016-09-01 01:59:59                                 1396.15   
2016-09-01 02:59:59                                 1398.08   
2016-09-01 03:59:59                                 1400.13   
2016-09-01 04:59:59                                 1402.17   

                     primary_cleaner.state.floatbank8_c_level  \
date                                                            
2016-09-01 00:59:59                                   -500.70   
2016-09-01 01:59:59           

primary_cleaner.input.sulfate  \
date                                                 
2017-12-31 19:59:59                         173.96   
2017-12-31 20:59:59                         172.91   
2017-12-31 21:59:59                         171.14   
2017-12-31 22:59:59                         179.70   
2017-12-31 23:59:59                         181.56   

                     primary_cleaner.input.depressant  \
date                                                    
2017-12-31 19:59:59                             15.96   
2017-12-31 20:59:59                             16.00   
2017-12-31 21:59:59                             15.99   
2017-12-31 22:59:59                             15.44   
2017-12-31 23:59:59                             15.00   

                     primary_cleaner.input.feed_size  \
date                                                   
2017-12-31 19:59:59                             8.07   
2017-12-31 20:59:59                             8.07   
2017-12-31 21:59:59                             8.07   
2017-12-31 22:59:59                             8.07   
2017-12-31 23:59:59                             8.07   

                     primary_cleaner.input.xanthate  \
date                                                  
2017-12-31 19:59:59                            0.90   
2017-12-31 20:59:59                            0.90   
2017-12-31 21:59:59                            1.17   
2017-12-31 22:59:59                            1.50   
2017-12-31 23:59:59                            1.62   

                     primary_cleaner.state.floatbank8_a_air  \
date                                                          
2017-12-31 19:59:59                                 1401.93   
2017-12-31 20:59:59                                 1447.08   
2017-12-31 21:59:59                                 1498.84   
2017-12-31 22:59:59                                 1498.47   
2017-12-31 23:59:59                                 1498.10   

                     primary_cleaner.state.floatbank8_a_level  \
date                                                            
2017-12-31 19:59:59                                   -499.73   
2017-12-31 20:59:59                                   -494.72   
2017-12-31 21:59:59                                   -501.77   
2017-12-31 22:59:59                                   -500.48   
2017-12-31 23:59:59                                   -499.80   

                     primary_cleaner.state.floatbank8_b_air  \
date                                                          
2017-12-31 19:59:59                                 1401.44   
2017-12-31 20:59:59                                 1448.85   
2017-12-31 21:59:59                                 1499.57   
2017-12-31 22:59:59                                 1497.99   
2017-12-31 23:59:59                                 1501.74   

                     primary_cleaner.state.floatbank8_b_level  \
date                                                            
2017-12-31 19:59:59                                   -499.19   
2017-12-31 20:59:59                                   -465.96   
2017-12-31 21:59:59                                   -495.52   
2017-12-31 22:59:59                                   -519.20   
2017-12-31 23:59:59                                   -505.15   

                     primary_cleaner.state.floatbank8_c_air  \
date                                                          
2017-12-31 19:59:59                                 1399.81   
2017-12-31 20:59:59                                 1443.89   
2017-12-31 21:59:59                                 1502.75   
2017-12-31 22:59:59                                 1496.57   
2017-12-31 23:59:59                                 1499.54   

                     primary_cleaner.state.floatbank8_c_level  \
date                                                            
2017-12-31 19:59:59                                   -499.60   
2017-12-31 20:59:59           


~~~~~~~~~~~~~~~~~~~~~~~ Основная описательная статистика ~~~~~~~~~~~~~~~~~~~~~~~~


primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                        5286.00                           5285.00   
mean                          174.84                              8.68   
std                            43.03                              3.07   
min                             2.57                              0.00   
25%                           147.12                              6.49   
50%                           177.83                              8.05   
75%                           208.13                             10.03   
max                           265.98                             40.00   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                          5290.00                         5286.00   
mean                              7.27                            1.38   
std                               0.61                            0.64   
min                               5.65                            0.00   
25%                               6.89                            0.91   
50%                               7.25                            1.20   
75%                               7.60                            1.80   
max                              15.50                            4.10   

       primary_cleaner.state.floatbank8_a_air  \
count                                 5290.00   
mean                                  1539.49   
std                                    116.80   
min                                      0.00   
25%                                   1498.94   
50%                                   1585.13   
75%                                   1602.08   
max                                   2103.10   

       primary_cleaner.state.floatbank8_a_level  \
count                                   5290.00   
mean                                    -497.67   
std                                       19.95   
min                                     -795.32   
25%                                     -500.36   
50%                                     -499.97   
75%                                     -499.57   
max                                      -57.20   

       primary_cleaner.state.floatbank8_b_air  \
count                                 5290.00   
mean                                  1545.17   
std                                    122.22   
min                                      0.00   
25%                                   1498.97   
50%                                   1595.62   
75%                                   1602.32   
max                                   1813.08   

       primary_cleaner.state.floatbank8_b_level  \
count                                   5290.00   
mean                                    -500.27   
std                                       32.97   
min                                     -800.00   
25%                                     -500.70   
50%                                     -500.03   
75%                                     -499.29   
max                                     -142.53   

       primary_cleaner.state.floatbank8_c_air  \
count                                 5290.00   
mean                                  1527.27   
std                                    122.54   
min                                      0.00   
25%                                   1473.23   
50%                                   1549.59   
75%                                   1601.14   
max                                   1715.05   

       primary_cleaner.state.floatbank8_c_level  \
count                                   5290.00   
mean                                    -498.33   
std                                       21.96   
min                                     -799.96   
25%                                     -501.02   
50%                                     -500.02   
75%                                     -498.99   
max                


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Тип данных ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 5290 entries, 2016-09-01 00:59:59 to 2017-12-31 23:59:59
Data columns (total 52 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   primary_cleaner.input.sulfate               5286 non-null   float64
 1   primary_cleaner.input.depressant            5285 non-null   float64
 2   primary_cleaner.input.feed_size             5290 non-null   float64
 3   primary_cleaner.input.xanthate              5286 non-null   float64
 4   primary_cleaner.state.floatbank8_a_air      5290 non-null   float64
 5   primary_cleaner.state.floatbank8_a_level    5290 non-null   float64
 6   primary_cleaner.state.floatbank8_b_air      5290 non-null   float64
 7   primary_cleaner.state.floatbank8_b_level    5290 non-null   float64
 8   primary_cleaner.state.floatbank8_c_air 

In [7]:
first_view(full)

~~~~~~~~~~~~~~~~~~ Обзор данных первых и последних пяти строк ~~~~~~~~~~~~~~~~~~~


final.output.concentrate_ag  final.output.concentrate_pb  \
date                                                                            
2016-01-15 00:00:00                         6.06                         9.89   
2016-01-15 01:00:00                         6.03                         9.97   
2016-01-15 02:00:00                         6.06                        10.21   
2016-01-15 03:00:00                         6.05                         9.98   
2016-01-15 04:00:00                         6.15                        10.14   

                     final.output.concentrate_sol  \
date                                                
2016-01-15 00:00:00                          5.51   
2016-01-15 01:00:00                          5.26   
2016-01-15 02:00:00                          5.38   
2016-01-15 03:00:00                          4.86   
2016-01-15 04:00:00                          4.94   

                     final.output.concentrate_au  final.output.recovery  \
date                                                                      
2016-01-15 00:00:00                        42.19                  70.54   
2016-01-15 01:00:00                        42.70                  69.27   
2016-01-15 02:00:00                        42.66                  68.12   
2016-01-15 03:00:00                        42.69                  68.35   
2016-01-15 04:00:00                        42.77                  66.93   

                     final.output.tail_ag  final.output.tail_pb  \
date                                                              
2016-01-15 00:00:00                 10.41                  0.90   
2016-01-15 01:00:00                 10.46                  0.93   
2016-01-15 02:00:00                 10.51                  0.95   
2016-01-15 03:00:00                 10.42                  0.88   
2016-01-15 04:00:00                 10.36                  0.79   

                     final.output.tail_sol  final.output.tail_au  \
date                                                               
2016-01-15 00:00:00                  16.90                  2.14   
2016-01-15 01:00:00                  16.63                  2.22   
2016-01-15 02:00:00                  16.21                  2.26   
2016-01-15 03:00:00                  16.53                  2.15   
2016-01-15 04:00:00                  16.53                  2.06   

                     primary_cleaner.input.sulfate  \
date                                                 
2016-01-15 00:00:00                         127.09   
2016-01-15 01:00:00                         125.63   
2016-01-15 02:00:00                         123.82   
2016-01-15 03:00:00                         122.27   
2016-01-15 04:00:00                         117.99   

                     primary_cleaner.input.depressant  \
date                                                    
2016-01-15 00:00:00                             10.13   
2016-01-15 01:00:00                             10.30   
2016-01-15 02:00:00                             11.32   
2016-01-15 03:00:00                             11.32   
2016-01-15 04:00:00                             11.91   

                     primary_cleaner.input.feed_size  \
date                                                   
2016-01-15 00:00:00                             7.25   
2016-01-15 01:00:00                             7.25   
2016-01-15 02:00:00                             7.25   
2016-01-15 03:00:00                             7.25   
2016-01-15 04:00:00                             7.25   

                     primary_cleaner.input.xanthate  \
date                                                  
2016-01-15 00:00:00                            0.99   
2016-01-15 01:00:00                            1.00   
2016-01-15 02:00:00                            0.99   
2016-01-15 03:00:00                            1.00   
2016-01-15 04:00:00                            1.01   

                     primary_cleaner.output.concen

final.output.concentrate_ag  final.output.concentrate_pb  \
date                                                                            
2018-08-18 06:59:59                         3.22                        11.36   
2018-08-18 07:59:59                         3.20                        11.35   
2018-08-18 08:59:59                         3.11                        11.43   
2018-08-18 09:59:59                         3.37                        11.63   
2018-08-18 10:59:59                         3.60                        11.74   

                     final.output.concentrate_sol  \
date                                                
2018-08-18 06:59:59                          6.80   
2018-08-18 07:59:59                          6.86   
2018-08-18 08:59:59                          6.89   
2018-08-18 09:59:59                          6.80   
2018-08-18 10:59:59                          6.72   

                     final.output.concentrate_au  final.output.recovery  \
date                                                                      
2018-08-18 06:59:59                        46.71                  73.76   
2018-08-18 07:59:59                        46.87                  69.05   
2018-08-18 08:59:59                        46.80                  67.00   
2018-08-18 09:59:59                        46.41                  65.52   
2018-08-18 10:59:59                        46.30                  70.28   

                     final.output.tail_ag  final.output.tail_pb  \
date                                                              
2018-08-18 06:59:59                  8.77                  3.14   
2018-08-18 07:59:59                  8.90                  3.13   
2018-08-18 08:59:59                  8.53                  2.91   
2018-08-18 09:59:59                  8.78                  2.82   
2018-08-18 10:59:59                  8.41                  2.52   

                     final.output.tail_sol  final.output.tail_au  \
date                                                               
2018-08-18 06:59:59                  10.40                  1.53   
2018-08-18 07:59:59                  10.55                  1.61   
2018-08-18 08:59:59                  11.12                  1.60   
2018-08-18 09:59:59                  10.46                  1.60   
2018-08-18 10:59:59                  10.65                  1.39   

                     primary_cleaner.input.sulfate  \
date                                                 
2018-08-18 06:59:59                         123.38   
2018-08-18 07:59:59                         120.88   
2018-08-18 08:59:59                         105.67   
2018-08-18 09:59:59                          98.88   
2018-08-18 10:59:59                          95.25   

                     primary_cleaner.input.depressant  \
date                                                    
2018-08-18 06:59:59                              8.03   
2018-08-18 07:59:59                              7.96   
2018-08-18 08:59:59                              7.96   
2018-08-18 09:59:59                              7.98   
2018-08-18 10:59:59                              8.08   

                     primary_cleaner.input.feed_size  \
date                                                   
2018-08-18 06:59:59                             6.50   
2018-08-18 07:59:59                             6.50   
2018-08-18 08:59:59                             6.50   
2018-08-18 09:59:59                             6.50   
2018-08-18 10:59:59                             6.50   

                     primary_cleaner.input.xanthate  \
date                                                  
2018-08-18 06:59:59                            1.30   
2018-08-18 07:59:59                            1.30   
2018-08-18 08:59:59                            1.32   
2018-08-18 09:59:59                            1.24   
2018-08-18 10:59:59                            1.28   

                     primary_cleaner.output.concen


~~~~~~~~~~~~~~~~~~~~~~~ Основная описательная статистика ~~~~~~~~~~~~~~~~~~~~~~~~


final.output.concentrate_ag  final.output.concentrate_pb  \
count                     19438.00                     19438.00   
mean                          5.17                         9.98   
std                           1.37                         1.67   
min                           0.00                         0.00   
25%                           4.25                         9.14   
50%                           5.07                        10.10   
75%                           5.90                        11.04   
max                          16.00                        17.03   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                      19228.00                     19439.00   
mean                           9.50                        44.08   
std                            2.79                         5.13   
min                            0.00                         0.00   
25%                            7.72                        43.40   
50%                            9.22                        45.01   
75%                           10.95                        46.28   
max                           19.62                        52.76   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count               19439.00              19438.00              19338.00   
mean                   67.05                  9.69                  2.71   
std                    10.13                  2.33                  0.95   
min                     0.00                  0.00                  0.00   
25%                    63.30                  8.06                  2.04   
50%                    68.17                  9.74                  2.75   
75%                    72.69                 11.13                  3.33   
max                   100.00                 19.55                  5.80   

       final.output.tail_sol  final.output.tail_au  \
count               19433.00              19439.00   
mean                   10.58                  3.04   
std                     2.87                  0.92   
min                     0.00                  0.00   
25%                     8.94                  2.46   
50%                    10.62                  2.98   
75%                    12.10                  3.57   
max                    22.32                  8.25   

       primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                       19415.00                          19402.00   
mean                          144.62                              8.82   
std                            44.46                              3.29   
min                             0.00                              0.00   
25%                           114.11                              6.08   
50%                           143.23                              8.04   
75%                           175.08                             11.01   
max                           265.98                             40.00   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                         19439.00                        19335.00   
mean                              7.31                            1.02   
std                               0.61                            0.51   
min                               1.08                            0.00   
25%                               6.94                            0.70   
50%                               7.28                            0.94   
75%                               7.67                            1.21   
max                              15.50                            4.10   

       primary_cleaner.output.concentrate_ag  \
count                               19439.00   
mean                                    8.44   
std                                     2.05   
min                                     0.00   
25%                                     7.26


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Тип данных ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 19439 entries, 2016-01-15 00:00:00 to 2018-08-18 10:59:59
Data columns (total 86 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   final.output.concentrate_ag                         19438 non-null  float64
 1   final.output.concentrate_pb                         19438 non-null  float64
 2   final.output.concentrate_sol                        19228 non-null  float64
 3   final.output.concentrate_au                         19439 non-null  float64
 4   final.output.recovery                               19439 non-null  float64
 5   final.output.tail_ag                                19438 non-null  float64
 6   final.output.tail_pb                                19338 non-null  float64
 7   final.output.tail_sol                  

In [8]:
def na_count(data):
    '''
    Функция принимает на вход датасет и выдаёт количество пропусков по столбцам, либо указывает что пропусков нет
    '''
    all_nulls = 0
    for col in data.columns:
        if data[col].isna().sum() > 0:
                all_nulls += 1
                print(data[col].isna().sum(), 'пропущенных значений в колонке', col, 'это', round(data[col].isna().mean() * 100, 1), '%')
    if all_nulls == 0:
        print('В датафрейме пропуски отсутствуют')

def doubles_count(data):
    '''
    Функция принимает на вход датасет и выдаёт количество дубликатов, либо указывает что их нет
    '''
    if data.duplicated().sum() > 0:
        print('Количество явных дубликатов - ', data.duplicated().sum())
    else:
        print('Явные дубликаты в датасете отсутствуют')

train.name = 'Тренировочная выборка'
test.name = 'Тестовая выборка'
full.name = 'Полный датасет'
datasets = [train, test, full]

In [9]:
# Проверим все датасеты на предмет дубликатов в цикле
for df in datasets:
    print('***', df.name, '***')
    doubles_count(df)
    print()

*** Тренировочная выборка ***
Явные дубликаты в датасете отсутствуют

*** Тестовая выборка ***
Явные дубликаты в датасете отсутствуют

*** Полный датасет ***
Явные дубликаты в датасете отсутствуют



In [10]:
# Проверим все датасеты на предмет пропусков в цикле
for df in datasets:
    print('***', df.name, '***')
    na_count(df)
    print()

*** Тренировочная выборка ***
1 пропущенных значений в колонке final.output.concentrate_ag это 0.0 %
1 пропущенных значений в колонке final.output.concentrate_pb это 0.0 %
211 пропущенных значений в колонке final.output.concentrate_sol это 1.5 %
100 пропущенных значений в колонке final.output.tail_pb это 0.7 %
5 пропущенных значений в колонке final.output.tail_sol это 0.0 %
20 пропущенных значений в колонке primary_cleaner.input.sulfate это 0.1 %
32 пропущенных значений в колонке primary_cleaner.input.depressant это 0.2 %
100 пропущенных значений в колонке primary_cleaner.input.xanthate это 0.7 %
86 пропущенных значений в колонке primary_cleaner.output.concentrate_pb это 0.6 %
286 пропущенных значений в колонке primary_cleaner.output.concentrate_sol это 2.0 %
1 пропущенных значений в колонке primary_cleaner.output.tail_ag это 0.0 %
15 пропущенных значений в колонке primary_cleaner.output.tail_pb это 0.1 %
46 пропущенных значений в колонке primary_cleaner.output.tail_sol это 0.3 %
4 про

Let's see the histograms

In [ ]:
full.iplot(kind='hist', title='Анализ гистограмм');
# Оценим гистограммы каждую по отдельности 
# Для анализа отдельных гистограмм по каждому столбцу, даблклик по названию столбца в списке справа (если показаны все)
# клик по следующему (если показан один) и клик по тому который необходимо убрать. Так анализируем каждый до конца.

Conclusions on the inspection of data:

- Not all features have a normal distribution.To build a model, you need to scaling, and you also need to get rid of emissions, many parameters have zero values, although they should not
- no duplicate
- many passes in different columns in different quantities

### We will check that the effectiveness of enrichment is calculated correctly

We check that the effectiveness of enrichment is calculated correctly.
On the training sample `train`, we calculate the effectiveness on your own and compare May with a sign` rougher.autPut.recovery`

In [11]:
C = train['rougher.output.concentrate_au']
F = train['rougher.input.feed_au']
T = train['rougher.output.tail_au']
rougher_recovery_calculation = C * (F - T) / (F * (C - T)) * 100

mae = mean_absolute_error(train['rougher.output.recovery'], rougher_recovery_calculation)

print(f'значение MAE = {mae:.17f}')

значение MAE = 0.00000000000000974


Mae is very small, the parameter is calculated for sure

### We will analyze the parameters inaccessible in the test sample

In [12]:
[col for col in list(train.columns) if (col not in list(test.columns))]

['final.output.concentrate_ag',
 'final.output.concentrate_pb',
 'final.output.concentrate_sol',
 'final.output.concentrate_au',
 'final.output.recovery',
 'final.output.tail_ag',
 'final.output.tail_pb',
 'final.output.tail_sol',
 'final.output.tail_au',
 'primary_cleaner.output.concentrate_ag',
 'primary_cleaner.output.concentrate_pb',
 'primary_cleaner.output.concentrate_sol',
 'primary_cleaner.output.concentrate_au',
 'primary_cleaner.output.tail_ag',
 'primary_cleaner.output.tail_pb',
 'primary_cleaner.output.tail_sol',
 'primary_cleaner.output.tail_au',
 'rougher.calculation.sulfate_to_au_concentrate',
 'rougher.calculation.floatbank10_sulfate_to_au_feed',
 'rougher.calculation.floatbank11_sulfate_to_au_feed',
 'rougher.calculation.au_pb_ratio',
 'rougher.output.concentrate_ag',
 'rougher.output.concentrate_pb',
 'rougher.output.concentrate_sol',
 'rougher.output.concentrate_au',
 'rougher.output.recovery',
 'rougher.output.tail_ag',
 'rougher.output.tail_pb',
 'rougher.output.ta

This list of features cannot be used for the model, since they are absent in a test sample, if we train the model using these features, the model when trying to make prediction in a test sample cannot find these features and will not work by giving an error

The list contains:
- calculated characteristics concentration of sulfates relative to gold in raw materials
- calculated characteristics of the raw materials of sulfates relative to gold in the fleet plants 10 and 11
- calculated characteristics of the ratio of gold and lead in raw materials
- dump tails of gold, silver, lead and salts at the flotation stage
- concentration of gold, silver, lead and salts at the flotation stage
- effectiveness of enrichment of black concentrate
- dump tails of gold, silver, lead and salts after primary cleaning
- concentration of gold, silver, lead and salts after primary cleaning
- dump tails of gold, silver, lead and salts after secondary cleaning
- final concentration of gold, silver, lead and salts
- final dump tails of gold, silver, lead and salts
- effectiveness of enrichment of the final concentrate

The effectiveness of enrichment is our target feature

### Bargaining

#### Names of the column

The speakers have very long names because of which it is not convenient to perceive the table, we will rename it right away

In [ ]:
dict_col_replaces = {'.': '_',
                     'output': 'out',
                     'final': 'fin',
                     'concentrate': 'conc',
                     'input': 'in',
                     'primary_cleaner': 'pr_cl',
                     'floatbank': 'fb',
                     'rougher': 'roug',
                     'secondary_cleaner': 'sc_cl',
                     'state': 'st',
                     'level': 'lvl',
                     'calculation': 'calc'
                     }

for i, j in dict_col_replaces.items():
    train.columns = train.columns.str.replace(i, j)
    test.columns = test.columns.str.replace(i, j)
    full.columns = full.columns.str.replace(i, j)

display(train.columns)
display(test.columns)
display(full.columns)

### Consider the correlations

It is necessary to take the signs for the model.As already indicated earlier, we need to weed out signs that are not in the test sample, now we will seize the signs that can interfere with us due to multicollinearity
- Consider the reign with targeted signs `fin_out_recovery` and` roug_out_recovery`.We use only columns that are available in the test sample and we need for the model
- Remove the signs that greatly correlate by other signs.These signs will not improve quality, but only complicate the model.

In [ ]:
col_for_corr = ['fin_out_recovery', 'roug_out_recovery'] + list(test.columns)

# train[col_for_corr].corr().iplot(kind="heatmap", title='Корреляции с целевыми признаками',
#                    colorscale='purples',
#                    dimensions=(1200,1200)
#                    )

The parameters of the floor in neighboring installations greatly correlate, for the model one value of completely correlating will be enough for the model

- `PR_CL_ST_FB8_A_AIR`,` PR_CL_ST_FB8_B_AIR` and `PR_CL_ST_FB8_C_AIR` Correlation 1, leave only A.B and de delete
- `PR_CL_ST_FB8_A_LVL`,` PR_CL_ST_FB8_B_LVL` and `PR_CL_ST_FB8_B_LVL` similarly to the previous point
- `PR_CL_ST_FB8_D_LVL, also greatly correlates with previous points, but not completely, so we will leave it
- `rous_in_fb10_xanthate` and` roug_in_fb1111_xanthate 'correlation 1, leave only the 10th installation
- `ROG_ST_FB10_B_LVL`,` ROG_ST_FB10_C_LVL`, `ROG_ST_FB10_D_LVL`,` ROG_ST_FB10_E_LVL`, `ROG_FB10_F_LVL` The correlation is very high, we leave only B. C, D, D, E and F delete
- `ROG_ST_FB10_B_AIR`,` ROG_ST_FB10_C_AIR`, `ROUG_ST_FB10_D_AIR`,` ROG_ST_FB10_E_AIR`, `ROG_FB10_F_AIR` similarly to the previous paragraph
- `SC_CL_ST_FB2_A_AIR` and` SC_CL_ST_FB2_B_AIR` The correlation is high, leave only A. B delete
- `SC_CL_ST_FB2_A_LVL` and` SC_CL_ST_FB2_B_LVL` similarly to the previous point

In [ ]:
col_to_drop = ['pr_cl_st_fb8_b_air',
               'pr_cl_st_fb8_c_air',
               'pr_cl_st_fb8_b_lvl',
               'pr_cl_st_fb8_c_lvl',
               'roug_in_fb11_xanthate',
               'roug_st_fb10_c_lvl',
               'roug_st_fb10_d_lvl',
               'roug_st_fb10_e_lvl',
               'roug_st_fb10_f_lvl',
               'roug_st_fb10_c_air', 
               'roug_st_fb10_d_air', 
               'roug_st_fb10_e_air', 
               'roug_st_fb10_f_air',
               'sc_cl_st_fb2_b_air',
               'sc_cl_st_fb2_b_lvl'
              ]

col_for_model = [col for col in list(test.columns) if col not in col_to_drop]

train = train.drop(col_to_drop, axis = 1)
test = test.drop(col_to_drop, axis = 1)

### ore passes

There are many passes.And in different columns.Perhaps there are lines in which passes are present in a significant number of parameters.If there are such, then we do not need them, to build a model, these lines will not be useful to us anyway.Let's try to see the lines from more than a third of the parameters with passes.

In [ ]:
full.loc[(full.isnull()).sum(axis=1) > full.shape[1] /3]

In [ ]:
train.loc[(train.isnull()).sum(axis=1) > train.shape[1] /3]

In [ ]:
test.loc[(test.isnull()).sum(axis=1) > test.shape[1] /3]

Remove these lines

In [ ]:
full = full.drop(full.loc[(full.isnull()).sum(axis=1) > full.shape[1] /2].index, axis = 0)
train = train.drop(train.loc[(train.isnull()).sum(axis=1) > train.shape[1] /2].index, axis = 0)

Once again, let's look at the passes in the columns of the model designed to build (gaps in columns that we do not use us very much for the model and worry)

In [ ]:
print('*** Тренировочная выборка ***')
na_count(train[col_for_model])
print()
print('*** Тестовая выборка ***')
na_count(test)

The remaining passes are a very small number of lines, no more than 3.1% to reduce the error of the model, it must be eliminated from passes.The task indicated that neighboring by date parameters are often similar, so we will fill in the passes with neighboring values

In [ ]:
train[col_for_model] = train[col_for_model].fillna(method='backfill')
test = test.fillna(method='backfill')

Interim conclusion on the block

- We loaded and studied the data;
- We checked that the calculation of efficiency was made correctly;
- Getting rid of passes
- deleted forabled columns
- removed columns that are absent in a test sample from building a model
- renamed the columns for the convenience of work

## data analysis

### Analysis of a change in the concentration of gold, silver and lead

In [ ]:
met_conc = train[['roug_out_conc_au',
            'pr_cl_out_conc_au',
            'fin_out_conc_au',
            'roug_out_conc_ag',
            'pr_cl_out_conc_ag',
            'fin_out_conc_ag',
            'roug_out_conc_pb',
            'pr_cl_out_conc_pb',
            'fin_out_conc_pb'
            ]]

met_conc.iplot(kind='box', title='Изменение концентрации металлов на разных этапах', xTitle='Дата', yTitle='Концентрация',
               colors=['#E4E5A0','#E5AF53','#E56723','#7EBDB2','#4BBD9D','#18BD78','#ABA8C3','#956BC3','#A94BC3']);

In [ ]:
met_conc.iplot(kind='hist', title='Изменение концентрации металлов на разных этапах', xTitle='Дата', yTitle='Концентрация',
               colors=['#E4E5A0','#E5AF53','#E56723','#7EBDB2','#4BBD9D','#18BD78','#ABA8C3','#956BC3','#A94BC3']);

In [ ]:
met_conc = train[['roug_out_conc_au',
            'pr_cl_out_conc_au',
            'fin_out_conc_au',
            'roug_out_conc_ag',
            'pr_cl_out_conc_ag',
            'fin_out_conc_ag',
            'roug_out_conc_pb',
            'pr_cl_out_conc_pb',
            'fin_out_conc_pb'
            ]].groupby(train.index.floor('D')).mean()

met_conc.iplot(title='Изменение концентрации металлов на разных этапах и со временем', xTitle='Дата', yTitle='Концентрация',
               colors=['#E4E5A0','#E5AF53','#E56723','#7EBDB2','#4BBD9D','#18BD78','#ABA8C3','#956BC3','#A94BC3']);

- The concentration of gold increases with each step in the cleaning, the concentration of silver decreases, the concentration of lead increases slightly after the index, then remains within the same level with a certain error.
- There are emissions up to zero values.
- values ​​are available in 2/3 years and for the last third (from September to the end of the year) are absent

### compare the distribution of granules sizes

In [ ]:
feeds_roug = pd.DataFrame({
    'train': train.loc[train['roug_in_feed_size'] < 150]['roug_in_feed_size'],
    'test': test.loc[test['roug_in_feed_size'] < 150]['roug_in_feed_size']
    })

fig, ax = plt.subplots(figsize=(13,6))
fig.suptitle('Распределение размера гранул на этапе флотации в выборках', fontsize=14, fontweight='bold')
sns.kdeplot(data=feeds_roug, fill=True, common_norm=False, alpha=0.4, ax=ax) 
ax.legend(['Тренировочная', 'Тестовая'])
ax.set_xlabel('Размер гранул')
ax.set_ylabel('Частота')
plt.show();

In [ ]:
feeds_pr_cl = pd.DataFrame({
    'train': train.loc[train['pr_cl_in_feed_size'] < 150]['pr_cl_in_feed_size'],
    'test': test.loc[test['pr_cl_in_feed_size'] < 150]['pr_cl_in_feed_size']
    })

fig, ax = plt.subplots(figsize=(13,6))
fig.suptitle('Распределение размера гранул после очистки в выборках', fontsize=14, fontweight='bold')
sns.kdeplot(data=feeds_pr_cl, fill=True, common_norm=False, alpha=0.4, ax=ax) 
ax.legend(['Тренировочная', 'Тестовая'])
ax.set_xlabel('Размер гранул')
ax.set_ylabel('Частота')
plt.show();

As can be seen on the graph, the distributions almost coincide, the differences are insignificant

### We examine the total concentration of all substances at different stages

In [ ]:
train['feed_conc'] = train['roug_in_feed_ag'] + train['roug_in_feed_pb'] + train['roug_in_feed_sol'] + train['roug_in_feed_au']
train['roug_conc'] = train['roug_out_conc_ag'] + train['roug_out_conc_pb'] + train['roug_out_conc_sol'] + train['roug_out_conc_au']
train['fin_conc'] = train['fin_out_conc_ag'] + train['fin_out_conc_pb'] + train['fin_out_conc_sol'] + train['fin_out_conc_au']

In [ ]:
met_sum_conc = train[['feed_conc','roug_conc','fin_conc']].groupby(train.index.floor('D')).mean()
met_sum_conc.iplot(title='Суммарная концентрация всех веществ на разных стадиях', xTitle='Дата', yTitle='Концентрация');

In some dates, the concentration falls to zero, let's see in histograms

In [ ]:
met_sum_conc = train[['feed_conc','roug_conc','fin_conc']]
met_sum_conc.iplot(kind='hist', title='Суммарная концентрация всех веществ на разных стадиях', xTitle='Дата', yTitle='Концентрация');

Conclusion

- the total concentration of substances increases after flotation
- after cleaning, the average value of the concentration of substances remains at the same level, but dispersion decreases and concentration becomes more predictable
- It is necessary to remove the lines with a zero concentration

In [ ]:
before_delete = len(train)
train = train.loc[(train['roug_in_feed_au'] > 0) &
                  (train['roug_in_feed_ag'] > 0) & 
                  (train['roug_in_feed_pb'] > 0) & 
                  (train['roug_in_feed_sol'] > 0) & 
                  (train['roug_out_conc_au'] > 0) & 
                  (train['roug_out_conc_ag'] > 0) &
                  (train['roug_out_conc_pb'] > 0) & 
                  (train['roug_out_conc_sol'] > 0) &
                  (train['fin_out_conc_au'] > 0) & 
                  (train['fin_out_conc_ag'] > 0) &
                  (train['fin_out_conc_pb'] > 0) & 
                  (train['fin_out_conc_sol'] > 0)
                 ]
print(f'Удалено {before_delete - len(train)} объектов')

Intermediate conclusion on the block:

- We studied the concentration of metals at different stages and concluded that the concentration of gold increases at each stage, silver is reduced, and the lead increases slightly after the location
- studied the distribution of granules in the samples and concluded that it coincides and will not spoil the results
- values ​​are available in 2/3 years and for the last third (from September to the end of the year) are absent
- the total concentration of substances increases after flotation
- after cleaning, the average value of the concentration of substances remains at the same level, but dispersion decreases and concentration becomes more predictable
- there are emissions up to zero values ​​that we deleted

## Model

### We will write a function for calculating the final Smape

Recall the formula from the description

$ {smape} = \ frac {1} {n} \ sum_ {i = 1}^{n} \ frac {| y_i - \ hat {y_i} {(| y_i | + | \ hat {y_i}}}|) / 2} * 100 \% $ $

$ $ {Final Smape} = 25 \% * {Smape (Rougher)} + ​​75 \% * {SMAPE (FINAL)} $ dollar

In [ ]:
def smape(y, y_pred):
    return 1 / len(y) * sum(abs(y - y_pred) / ((abs(y) + abs(y_pred)) / 2)) * 100

def fin_smape(roug_smape, fin_smape):
    return 0.25 * roug_smape + 0.75 * fin_smape

### Learning and Assessment of the quality of models

#### We will divide the dataset to sample

In [ ]:
# Для построения модели предсказания эффективности на этапе флотации, 
# нам необходимо использовать только параметры связанные с этим этапом. 
# Сделаем список этих параметров
col_for_roug_model = [col for col in train[col_for_model].columns if col.startswith('roug')]

roug_train_x = train[col_for_roug_model]
roug_train_y = train['roug_out_recovery']
roug_test_x = test[col_for_roug_model]
roug_test_y = full.loc[test.index, 'roug_out_recovery']

fin_train_x = train[col_for_model]
fin_train_y = train['fin_out_recovery']
fin_test_x = test
fin_test_y = full.loc[test.index, 'fin_out_recovery']

if set(roug_train_x) != set(roug_test_x):
    print('Ошибка! Признаки в тренировочной и тестовой выборках для модели рассчета эффективности на этапе флотации не совпадают')
elif set(fin_train_x) != set(fin_test_x):
    print('Ошибка! Признаки в тренировочной и тестовой выборках для модели рассчета эффективности не совпадают')
else:
    print('Признаки в тренировочной и тестовой выборках совпадают')

In [ ]:
roug_train_x.iplot(kind='hist', title='проверка параметров перед обучением модели', xTitle='Значение', yTitle='Концентрация');

In [ ]:
fin_train_x.iplot(kind='hist', title='проверка параметров перед обучением модели', xTitle='Значение', yTitle='Концентрация');

### Create Kfold Object

We build different models and we want to be sure that each time we work with the equally broken data during a cross.To do this, create a kfold object

In [ ]:
kf = KFold(n_splits=5, random_state=1337, shuffle=True)

### Create Scoring Object to use it in ross validation

In [ ]:
scor = make_scorer(smape, greater_is_better=False)

#### linear regression

In [ ]:
# Масштабируем фичи
roug_scaler = StandardScaler()
fin_scaler = StandardScaler()
roug_scaler.fit(roug_train_x)
fin_scaler.fit(fin_train_x)

roug_train_x_scaled = pd.DataFrame(roug_scaler.transform(roug_train_x), columns=col_for_roug_model)
fin_train_x_scaled = pd.DataFrame(fin_scaler.transform(fin_train_x), columns=col_for_model)
roug_test_x_scaled = pd.DataFrame(roug_scaler.transform(roug_test_x), columns=col_for_roug_model)
fin_test_x_scaled = pd.DataFrame(fin_scaler.transform(fin_test_x), columns=col_for_model)

# обучим модель и проверим её кросс валидацией
lr_roug_model = LinearRegression()
lr_fin_model = LinearRegression()

lr_roug_smape = cross_val_score(lr_roug_model,
                                roug_train_x_scaled, 
                                roug_train_y, 
                                scoring = scor,
                                cv = kf,
                                n_jobs = -1).mean()

lr_fin_smape = cross_val_score(lr_fin_model, 
                               fin_train_x_scaled,
                               fin_train_y, 
                               scoring = scor,
                               cv = kf,
                               n_jobs = -1).mean()

lr_final_smape = fin_smape(abs(lr_roug_smape), abs(lr_fin_smape))

print('Линейная регрессия:')
print(f'sMAPE на этапе флотации = {round(abs(lr_roug_smape), 2)}')
print(f'sMAPE на финальном этапе = {round(abs(lr_fin_smape), 2)}')
print(f'Final sMAPE = {round(lr_final_smape, 2)}')

### Casual Forest

To select parameters, we use the RandomizedSearchcv method

In [ ]:
rf_roug_model = RandomForestRegressor()

rf_params = {'random_state': [1337],
             'n_estimators': range(100, 1001, 100),
             'max_depth': [1,4,8,12,16,20],
             'min_samples_split' : [2,6,10,16],
             'min_samples_leaf' : [1,2,4,8,16]
             }

rf_roug_rndsch = RandomizedSearchCV(rf_roug_model,
                                   param_distributions = rf_params,
                                   n_iter = 500,
                                   scoring = scor,
                                   cv = kf,
                                   random_state = 1337,
                                   n_jobs = -1
                                   )

rf_roug_rndsch.fit(roug_train_x, roug_train_y)

rf_roug_smape = abs(rf_roug_rndsch.best_score_)

print('Лучшее значение sMAPE модели случайного леса на этапе подготовки сырья к флотации:')
print(rf_roug_smape)
print('при следующих параметрах:')
print(rf_roug_rndsch.best_params_)

In [ ]:
Лучшее значение sMAPE модели случайного леса на этапе подготовки сырья к флотации:
2.7113450337295015
при следующих параметрах:
{'random_state': 1337, 'n_estimators': 1000, 'min_samples_split': 2, 'min_samples_leaf': 1, 'max_depth': 20}

# rf_roug_smape = 2.7113450337295015

In [ ]:
rf_fin_model = RandomForestRegressor()

rf_fin_rndsch = RandomizedSearchCV(rf_fin_model,
                                   param_distributions = rf_params,
                                   n_iter = 500,
                                   scoring = scor,
                                   cv = kf,
                                   random_state = 1337,
                                   n_jobs = -1
                                   )

rf_fin_rndsch.fit(fin_train_x, fin_train_y)

rf_fin_smape = abs(rf_fin_rndsch.best_score_)

print('Лучшее значение sMAPE модели случайного леса на этапе подготовки сырья к флотации:')
print(rf_fin_smape)
print('при следующих параметрах:')
print(rf_fin_rndsch.best_params_)

In [ ]:
# Лучшее значение sMAPE модели случайного леса на этапе подготовки сырья к флотации:
# 5.524318574514315
# при следующих параметрах:
# {'random_state': 1337, 'n_estimators': 900, 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_depth': 20}

# rf_fin_smape = 5.524318574514315

In [ ]:
rf_final_smape = fin_smape(rf_roug_smape, rf_fin_smape)

print('Случайный Лес:')
print(f'sMAPE на этапе флотации = {round(rf_roug_smape, 2)}')
print(f'sMAPE на финальном этапе = {round(rf_fin_smape, 2)}')
print(f'Final sMAPE = {round(rf_final_smape, 2)}')

### compare the models

In [ ]:
model_names = ['Linear Regression', 
               'Random Forest',  
               ]
             
final_smape = [lr_final_smape,
               rf_final_smape,
               ]

models_dif = pd.DataFrame({'Модель':model_names,
                           'final sMAPE score':final_smape
                           })
models_dif

A random forest model showed the best result on Smape metric

#### Testing the best model

In [ ]:
rf_roug_model = RandomForestRegressor(random_state=1337, n_estimators=1000, min_samples_split=2, min_samples_leaf=1, max_depth=20)
rf_fin_model = RandomForestRegressor(random_state=1337, n_estimators=900, min_samples_split=2, min_samples_leaf=2, max_depth=20)

rf_roug_model.fit(roug_train_x, roug_train_y)
rf_fin_model.fit(fin_train_x, fin_train_y)

rf_roug_pred = rf_roug_model.predict(roug_test_x)
rf_fin_pred = rf_fin_model.predict(fin_test_x)

test_rough_smape = smape(roug_test_y, rf_roug_pred)
test_fin_smape = smape(fin_test_y, rf_fin_pred)
test_final_smape = fin_smape(test_rough_smape, test_fin_smape)

print('Случайный лес - Тестовая выборка:')
print(f'sMAPE на этапе флотации = {round(test_rough_smape, 2)}')
print(f'sMAPE на финальном этапе = {round(test_fin_smape, 2)}')
print(f'Final sMAPE = {round(test_final_smape, 2)}')

### compare with the constant dummy model

In [ ]:
dummy_roug_model = DummyRegressor(strategy="median")
dummy_fin_model = DummyRegressor(strategy="median")

dummy_roug_model.fit(roug_train_x, roug_train_y)
dummy_fin_model.fit(fin_train_x, fin_train_y)

dummy_roug_pred = dummy_roug_model.predict(roug_test_x)
dummy_fin_pred = dummy_fin_model.predict(fin_test_x)

dummy_roug_smape = smape(roug_test_y, dummy_roug_pred)
dummy_fin_smape = smape(fin_test_y, dummy_fin_pred)

dummy_final_smape = fin_smape(dummy_roug_smape, dummy_fin_smape)

print('Константная dummy регрессия:')
print(f'sMAPE на этапе флотации = {round(dummy_roug_smape, 2)}')
print(f'sMAPE на финальном этапе = {round(dummy_fin_smape, 2)}')
print(f'Final sMAPE = {round(dummy_final_smape, 2)}')

## Conclusion

Work done:
- data loaded and studied
- in the data, the passes are found and removed
- Calculations of the effectiveness of enrichment in the data are verified - compose obtained by the formula
- The signs of a test sample are analyzed and signs that are missing in it are not used in teaching models in the training sample
- the concentration of metals at different stages has been investigated, conclusions were drawn that the concentration of gold grows with each stage, silver falls, and lead grows a little after flotation
- also at this stage an abnormal data in which the concentration of metals is null - these lines are deleted
- the size of the granules in the training and test samples is investigated - the distribution coincides
- Functions for calculating the metrics SMAPE are written
- To search for the best model, they used K-Fold Cross Validation
- The linear regression model is trained.Metric SMAPE for ross validation:
- Smape at the flotation stage = 5.18
- SMAPE at the final stage = 8.03
- Final Smape = 7.32
- The model of random forest is trained.Metric SMAPE for ross validation:
- Smape at the flotation stage = 2.71
- SMAPE at the final stage = 5.52
- Final Smape = 4.82
- The best model of the trained was a model of random forest.Smape metric in a test sample:
- Smape at the flotation stage = 8.02
- Smape at the final stage = 9.54
- Final Smape = 9.16
- The Constant Dummy Model is trained to check our model for adequacy.Her Metric Smape in a test sample:
- Smape at the flotation stage = 8.48
- SMAPE at the final stage = 9.69
- Final Smape = 9.38

Our model shows SMAPE better than Dummy model at 0.22